Add ChainRule for `keyless_unname`

glennmoy opened the pull request:
Essentially pair-programmed with @mzgubic.

Needed to add an `rrule` for `keyless_unname` for a downstream AD use-case.

mzgubic commented on Project.toml:
```diff
@@ -19,7 +20,10 @@ Tables = "bd369af6-aec1-5ad0-b16a-f7cc5008161c"
+ChainRulesCore = "1"
+ChainRulesTestUtils = "1"
 CovarianceEstimation = "0.2"
+FiniteDifferences = "0.12"
```
these and CRTU should go

mzgubic replied:
Ah sorry, my bad. For some reason I saw it in the deps column.

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,22 @@
+
+function ChainRulesCore.rrule(::typeof(keyless_unname), x::KaNda)
+    project_x = ProjectTo(x.data)
+    pb(y) = _KeyedArray_pullback(project_x(y), named_axiskeys(x))
```
can we rename this and the other similar ones to 
```suggestion
    keyless_unname_pb(y) = _KeyedArray_pullback(project_x(y), named_axiskeys(x))
```

mzgubic commented on test/runtests.jl:
```diff
@@ -1,5 +1,7 @@
 using Test, AxisKeys, NamedDims
 using Statistics, OffsetArrays, Tables, UniqueVectors, LazyStack
+using ChainRulesTestUtils: test_rrule
```
```suggestion
using ChainRulesTestUtils
```

mzgubic replied:
I guess it's just style, and since `test_rrule` is the main functionality of CRTU I guess it doesn't need to be told where it's from? Idk, don't feel strongly about this

mcabbott commented:
> orthogonal to the changes I am adding here

I should look closer but `_KeyedArray_pullback(ȳ::AbstractArray, keys) = (NoTangent(), wrapdims(ȳ, keys...))` looks like it wants to attach key vectors to the gradient. I'm not sure that's a good idea. Their values are part of the original `x`, not its gradient. Can't it just stay an array?

glennmoy commented:
> > orthogonal to the changes I am adding here
> 
> I should look closer but `_KeyedArray_pullback(ȳ::AbstractArray, keys) = (NoTangent(), wrapdims(ȳ, keys...))` looks like it wants to attach key vectors to the gradient. I'm not sure that's a good idea. Their values are part of the original `x`, not its gradient. Can't it just stay an array?

Ah I see, yeah that makes sense. I've changed it to return an array and refactored the code to be more simple.

oxinabox commented on src/chainrules.jl:
```diff
@@ -0,0 +1,15 @@
+_keyless_unname_pullback(ȳ::AbstractThunk) = _keyless_unname_pullback(unthunk(ȳ))
+
+function ChainRulesCore.rrule(::typeof(keyless_unname), x::KeyedArray)
+    project_x = ProjectTo(x.data)
```
I think we should actually define a `ProjectTo(x::KeyedArray)` that returns a `ProjectTo{KeyedArray}` which not only projects the backing data, but also reattaches the keys from the primal to the tangent.

glennmoy commented:
bump

oxinabox commented on test/_chainrules.jl:
```diff
@@ -0,0 +1,34 @@
+@testset "chainrules.jl" begin
+
+    function FiniteDifferences.to_vec(k::AxisKeys.KaNda)
+        v, b = to_vec(k.data)
+        back(x) = wrapdims(b(x); AxisKeys.named_axiskeys(k)...)
+        return v, back
+    end
+
+    function FiniteDifferences.to_vec(k::KeyedArray)
+        v, b = to_vec(k.data)
+        back(x) = wrapdims(b(x), axiskeys(k)...)
+        return v, back
```
Do we actually need these `to_vec` definitions?
FiniteDifferences.jl uses ProjectTo under-the-hood

glennmoy replied:
yes, I think so. Without them all of the tests fail.

mzgubic replied:
Have you tested this after adding `ProjectTo`? It might just work now.

glennmoy replied:
yes, I removed the `to_vec` functions, tested and they still fail. But not for the same reason.

>FiniteDifferences.jl uses ProjectTo under-the-hood

Where does it use it? I went looking but didn't find it.

oxinabox replied:
my mistake, it is used in `ChainRulesTestUtils.rand_tangent`

oxinabox replied:
And yeah, looking at FiniteDifferences i see that indeed right now it doesn't handle `AbstractArray`s in general.
Except by handling them as structs, which would make `Tangent`s.

glennmoy commented:
bump

oxinabox commented on src/chainrules.jl:
```diff
@@ -0,0 +1,20 @@
+    return ProjectTo{KeyedArray}(;data=ProjectTo(keyless(x)), keys=axiskeys(x))
+end
+
+(project::ProjectTo{KeyedArray})(dx) = wrapdims(project.data(dx), project.keys...)
```
These need some direct tests of there own I think.

Testing that the keys of the result of `ProjectTo(ka)(dk)` for some plain `dk::Matrix`  and some `KeyedArray` `ka`, match the keys of `ka`

glennmoy replied:
added some unit tests

oxinabox commented:
@glennmoy  tests are failing

glennmoy commented:
> 

fixed

glennmoy commented:
bump
